File: examples/nostr-basics/04-relay-bridge/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/gorilla/websocket"
)

type readResult struct {
	data []byte
	err  error
}

func readMessages(conn *websocket.Conn, output chan<- readResult) {
	for {
		_, data, err := conn.ReadMessage()
		output <- readResult{data: data, err: err}
		if err != nil {
			return
		}
	}
}

func bridgeKinds() []int {
	value := os.Getenv("NOSTR_BRIDGE_KINDS")
	if value == "" {
		return []int{1}
	}
	var result []int
	for _, part := range strings.Split(value, ",") {
		var kind int
		if _, err := fmt.Sscanf(strings.TrimSpace(part), "%d", &kind); err == nil {
			result = append(result, kind)
		}
	}
	if len(result) == 0 {
		return []int{1}
	}
	return result
}

func main() {
	if len(os.Args) < 3 {
		panic("usage: nostr-04-relay-bridge <source-ws-url> <sink-ws-url>")
	}
	sourceURL, sinkURL := os.Args[1], os.Args[2]
	if sourceURL == sinkURL {
		panic("source and sink must be different relay URLs")
	}

	source, _, err := websocket.DefaultDialer.Dial(sourceURL, nil)
	if err != nil {
		panic(err)
	}
	defer source.Close()
	sink, _, err := websocket.DefaultDialer.Dial(sinkURL, nil)
	if err != nil {
		panic(err)
	}
	defer sink.Close()

	const subscriptionID = "bridge-source"
	filter := map[string]any{"kinds": bridgeKinds()}
	if err := source.WriteJSON([]any{"REQ", subscriptionID, filter}); err != nil {
		panic(err)
	}
	fmt.Printf("subscribed to %s: %v\nforwarding signed events to %s\n", sourceURL, filter, sinkURL)

	sourceMessages := make(chan readResult)
	sinkMessages := make(chan readResult)
	go readMessages(source, sourceMessages)
	go readMessages(sink, sinkMessages)

	for {
		select {
		case result := <-sourceMessages:
			if result.err != nil {
				fmt.Println("source closed:", result.err)
				_ = source.WriteJSON([]any{"CLOSE", subscriptionID})
				return
			}
			var envelope []json.RawMessage
			if err := json.Unmarshal(result.data, &envelope); err != nil || len(envelope) == 0 {
				continue
			}
			var label string
			_ = json.Unmarshal(envelope[0], &label)
			switch label {
			case "EVENT":
				if len(envelope) < 3 {
					continue
				}
				var subID string
				_ = json.Unmarshal(envelope[1], &subID)
				if subID != subscriptionID {
					continue
				}
				var eventMeta struct {
					ID string `json:"id"`
				}
				if err := json.Unmarshal(envelope[2], &eventMeta); err != nil {
					panic(err)
				}
				// Keep the received event object byte-for-byte intact. The bridge
				// forwards the original signature; it does not rebuild or re-sign it.
				if err := sink.WriteJSON([]any{"EVENT", json.RawMessage(envelope[2])}); err != nil {
					panic(err)
				}
				fmt.Println("forwarded EVENT", eventMeta.ID)
			case "EOSE":
				fmt.Println("source EOSE: historical events are complete")
			case "NOTICE", "CLOSED":
				fmt.Println("source", label, string(result.data))
			}
		case result := <-sinkMessages:
			if result.err != nil {
				fmt.Println("sink closed:", result.err)
				return
			}
			var envelope []json.RawMessage
			if err := json.Unmarshal(result.data, &envelope); err != nil || len(envelope) == 0 {
				continue
			}
			var label string
			_ = json.Unmarshal(envelope[0], &label)
			if label == "OK" || label == "NOTICE" || label == "CLOSED" {
				fmt.Println("sink", label, string(result.data))
			}
		}
	}
}

File: examples/nostr-basics/04-relay-bridge/rust/src/main.rs
use std::error::Error;

use futures_util::{SinkExt, StreamExt};
use nostr::Event;
use serde_json::{json, Value};
use tokio_tungstenite::{connect_async, tungstenite::Message};

const SUBSCRIPTION_ID: &str = "bridge-source";

fn kinds() -> Vec<u16> {
    std::env::var("NOSTR_BRIDGE_KINDS")
        .ok()
        .map(|value| {
            value
                .split(',')
                .filter_map(|part| part.trim().parse::<u16>().ok())
                .collect()
        })
        .filter(|values: &Vec<u16>| !values.is_empty())
        .unwrap_or_else(|| vec![1])
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let source_url = std::env::args()
        .nth(1)
        .ok_or("usage: nostr-04-relay-bridge <source-ws-url> <sink-ws-url>")?;
    let sink_url = std::env::args()
        .nth(2)
        .ok_or("usage: nostr-04-relay-bridge <source-ws-url> <sink-ws-url>")?;
    if source_url == sink_url {
        return Err("source and sink must be different relay URLs".into());
    }

    let (source, _) = connect_async(&source_url).await?;
    let (sink, _) = connect_async(&sink_url).await?;
    let (mut source_write, mut source_read) = source.split();
    let (mut sink_write, mut sink_read) = sink.split();

    let filter = json!({"kinds": kinds()});
    source_write
        .send(Message::Text(
            json!(["REQ", SUBSCRIPTION_ID, filter]).to_string().into(),
        ))
        .await?;
    println!("subscribed to {source_url}: {filter}");
    println!("forwarding signed events to {sink_url}");

    loop {
        tokio::select! {
            source_message = source_read.next() => {
                let Some(message) = source_message else { break; };
                let Message::Text(text) = message? else { continue; };
                let value: Value = serde_json::from_str(text.as_ref())?;
                let Some(array) = value.as_array() else { continue; };
                match array.first().and_then(Value::as_str) {
                    Some("EVENT") if array.get(1).and_then(Value::as_str) == Some(SUBSCRIPTION_ID) => {
                        let raw_event = array.get(2).ok_or("source EVENT has no event payload")?;
                        let event: Event = serde_json::from_value(raw_event.clone())?;
                        let id = event.id.to_hex();
                        sink_write.send(Message::Text(
                            json!(["EVENT", event]).to_string().into(),
                        )).await?;
                        println!("forwarded EVENT {id}");
                    }
                    Some("EOSE") => println!("source EOSE: historical events are complete"),
                    Some("NOTICE") => println!("source NOTICE: {value}"),
                    Some("CLOSED") => println!("source CLOSED: {value}"),
                    _ => {}
                }
            }
            sink_message = sink_read.next() => {
                let Some(message) = sink_message else { break; };
                let Message::Text(text) = message? else { continue; };
                let value: Value = serde_json::from_str(text.as_ref())?;
                match value.as_array().and_then(|array| array.first()).and_then(Value::as_str) {
                    Some("OK") | Some("NOTICE") | Some("CLOSED") => println!("sink: {value}"),
                    _ => {}
                }
            }
        }
    }

    source_write
        .send(Message::Text(
            json!(["CLOSE", SUBSCRIPTION_ID]).to_string().into(),
        ))
        .await?;
    Ok(())
}
